apps/web/public/sw.js

```diff
@@ -0,0 +1,411 @@
+/* Service Worker */
+
+const version = 1;
+const dbVersion = 2;
+const database = "inbox-zero";
+let DB = null;
+let dynamicName = `dynamicCache`;
+let IMAGE_CACHE = "IMAGE_CACHE";
+const LABEL_STORE = {
+  name: "labels",
+  key: "id",
+  indexes: [],
+};
+const EMAIL_STORE = {
+  name: "emails",
+  key: "gmailMessageId",
+  indexes: [{ name: "timestampIDX", property: "timestamp", params: {} }],
+};
+const ALL_STORES = [LABEL_STORE, EMAIL_STORE];
+
+const LABELS_UPDATED = "LABELS_UPDATED";
+const MAILS_UPDATED = "MAILS_UPDATED";
+
+const MESSAGES = { LABELS_UPDATED, MAILS_UPDATED };
+
+const TIME_CONSTANT = {
+  // in ms
+  hour: 3600_000,
+  day: 86400_000,
+  week: 604800_000,
+  month: 2629743_000,
+  year: 31556926_000,
+};
+
+self.addEventListener("install", (event) => {
+  console.log(`Version ${version} installed`);
+  self.skipWaiting();
+});
+
+self.addEventListener("activate", (event) => {
+  //service worker is activated
+  event.waitUntil(Promise.resolve().then(openDB));
+  console.log("activated");
+});
+
+self.addEventListener("fetch", (event) => {
+  //service worker intercepted a fetch call
+  console.log(event.request.url);
+  if (!isAuth(event.request.url))
+    event.respondWith(
+      handleFetch(event.request).then((response) => {
+        // TODO: Remove this
+        if (event.request.url.includes("tinybird")) {
+          console.log(response);
+        }
+        return response;
+      }),
+    );
+});
+
+self.addEventListener("message", (event) => {
+  //message from webpage
+});
+
+const openDB = (callback) => {
+  return new Promise((resolve, reject) => {
+    let req = indexedDB.open(database, dbVersion);
+    req.onerror = (err) => {
+      //could not open db
+      console.warn(err);
+      DB = null;
+      reject(err);
+    };
+    req.onupgradeneeded = (ev) => {
+      let db = ev.target.result;
+      for (const store of ALL_STORES) {
+        const { name, key, indexes } = store;
+        if (!db.objectStoreNames.contains(name)) {
+          const objectStore = db.createObjectStore(name, {
+            keyPath: key,
+          });
+
+          for (const index of indexes) {
+            const { name, property, params } = index;
+            objectStore.createIndex(name, property, params);
+          }
+        }
+      }
+    };
+    req.onsuccess = (ev) => {
+      DB = ev.target.result;
+      console.log("db opened and upgraded as needed");
+      if (callback) {
+        callback();
+      }
+      resolve();
+    };
+  });
+};
+
+async function handleFetch(request) {
+  const { url } = request;
+  if (url.includes("/api/google/labels")) {
+    // this should be called regardless.
+    fetchAndUpdateIDB(request.clone());
```

Comment:
This function fetches new/updated labels from server in the background and updates indexeddb and client with them.